L1Trigger/TrackerTFP/src/HoughTransform.cc

```diff
@@ -0,0 +1,200 @@
+      acceptedAll.erase(limit, acceptedAll.end());
+      // store found tracks
+      auto put = [](const deque<StubHT*>& stubs, TTDTC::Stream& stream){
+        stream.reserve(stubs.size());
+        for (StubHT* stub : stubs)
+          stream.emplace_back(stub ? stub->frame() : TTDTC::Frame());
+      };
+      const int offset = region_ * dataFormats_->numChannel(Process::ht);
+      put(acceptedAll, accepted[offset + binInv2R]);
+      // store lost tracks
+      put(lostAll, lost[offset + binInv2R]);
+    }
+  }
+
+  // associate stubs with inv2R and phiT bins
+  void HoughTransform::fillIn(deque<StubGP*>& inputSector, vector<StubHT*>& acceptedSector, vector<StubHT*>& lostSector, int inv2R) {
+    // fifo, used to store stubs which belongs to a second possible track
+    deque<StubHT*> stack;
+    // clock accurate firmware emulation, each while trip describes one clock tick, one stub in and one stub out per tick
+    while (!inputSector.empty() || !stack.empty()) {
+      StubHT* stubHT = nullptr;
+      StubGP* stubGP = pop_front(inputSector);
+      if (stubGP) {
+        const double phiT = stubGP->phi() - inv2R_.floating(inv2R) * stubGP->r();
+        const int major = phiT_.integer(phiT);
+        if (phiT_.inRange(major)) {
+          // major candidate has pt > threshold (3 GeV)
+          stubsHT_.emplace_back(*stubGP, major, inv2R);
+          stubHT = &stubsHT_.back();
+        }
+        const double chi = phiT - phiT_.floating(major);
+        if (abs(stubGP->r() * inv2R_.base()) + 2. * abs(chi) >= phiT_.base()) {
+          // stub belongs to two candidates
+          const int minor = chi >= 0. ? major + 1 : major - 1;
+          if (phiT_.inRange(minor)) {
+            // second (minor) candidate has pt > threshold (3 GeV)
+            stubsHT_.emplace_back(*stubGP, minor, inv2R);
+            if (enableTruncation_ && (int)stack.size() == setup_->htDepthMemory() - 1)
+              // buffer overflow
+              lostSector.push_back(pop_front(stack));
+            // store minor stub in fifo
+            stack.push_back(&stubsHT_.back());
+          }
+        }
+      }
+      // take a minor stub if no major stub available
+      acceptedSector.push_back(stubHT ? stubHT : pop_front(stack));
+    }
+    // truncate to many input stubs
+    const auto limit = enableTruncation_ ? next(acceptedSector.begin(), min(setup_->numFrames(), (int)acceptedSector.size())) : acceptedSector.end();
+    copy_if(limit, acceptedSector.end(), back_inserter(lostSector), [](StubHT* stub){ return stub; });
+    acceptedSector.erase(limit, acceptedSector.end());
+  }
+
+  // identify tracks
+  void HoughTransform::readOut(const vector<StubHT*>& acceptedSector, const vector<StubHT*>& lostSector, deque<StubHT*>& acceptedAll, deque<StubHT*>& lostAll) const {
+    // used to recognise in which order tracks are found
+    TTBV patternPhiTs(0, setup_->htNumBinsPhiT());
```

Comment:
Rename "patternPhiTs" to "trkFoundPhiTs", to make it clearer that this array identifies cells where a track has already been found, and to make clear that it is not something similar to array "patternHits".

Reply:
done